Donation-platform/src/test/java/com/google/PersonalizedRecommendationsTest.java

```diff
@@ -11,269 +11,340 @@
-
+ 
     Assert.assertTrue(everyCharityPresent && correctOrdering);
   }
-
+ 
   @Test
   public void onlyTags2and3Selected() {
     // If only tags #2 and #3 are selected, recommend all charities that have tag
     // #3, tag #3, or both #2 and #3. Every charity that has tag #2 should come before
     // every charity that does not have tag #2.
-
+ 
     // Example: All charities with the tags "health" and/or "children" should be
     // recommended. Every charity with the tag "health" should come before every
     // charity without the tag "health."
-
+ 
     boolean everyCharityPresent = false;
     boolean correctOrdering = false;
-
+ 
     List<String> tags = Arrays.asList(null, "health", "children");
     List<Charity> actual = recommendation.getBestMatches(tags);
     List<Charity> expected = Arrays.asList(SJ, RC, AHA, YMCA);
-    everyCharityPresent = (new HashSet<>(expected)).equals(new HashSet<>(actual));
-
+    everyCharityPresent = sameSetOfCharities(actual, expected);
+ 
     // Find the index of the highest-ranking charity without tag #2
     int indexOfHighestRankingCharityWithoutTag2 = actual.size() - 1;
     for (Charity charity : actual) {
-      if (!(new HashSet<>(charity.getTags())).contains("health")
+      Collection<Tag> charityTagObjects = charity.getCategories();
+      Collection<String> charityTagNames = new ArrayList<String>();
+      for(Tag tag : charityTagObjects) {
+        charityTagNames.add(tag.getName());
+      }
+      if (!(new HashSet<>(charityTagNames)).contains("health")
           && actual.indexOf(charity) < indexOfHighestRankingCharityWithoutTag2) {
         indexOfHighestRankingCharityWithoutTag2 = actual.indexOf(charity);
       }
     }
     // Find the index of the lowest-ranking charity with tag #2
     int indexOfLowestRankingCharityWithTag2 = 0;
     for (Charity charity : actual) {
-      if ((new HashSet<>(charity.getTags())).contains("health")
+      Collection<Tag> charityTagObjects = charity.getCategories();
+      Collection<String> charityTagNames = new ArrayList<String>();
+      for(Tag tag : charityTagObjects) {
+        charityTagNames.add(tag.getName());
+      }
+      if ((new HashSet<>(charityTagNames)).contains("health")
           && actual.indexOf(charity) > indexOfLowestRankingCharityWithTag2) {
         indexOfLowestRankingCharityWithTag2 = actual.indexOf(charity);
       }
     }
     correctOrdering = indexOfLowestRankingCharityWithTag2 < indexOfHighestRankingCharityWithoutTag2;
-
+ 
     Assert.assertTrue(everyCharityPresent && correctOrdering);
   }
-}
+
+  /* Checks if two charity collections have identical sets of charities (order does not matter). */
+  public static boolean sameSetOfCharities(Collection<Charity> charities1, Collection<Charity> charities2){
+    HashSet<String> charity1Names = new HashSet<String>();
+    HashSet<String> charity2Names = new HashSet<String>();
+    for(Charity charity : charities1) {
+      charity1Names.add(charity.getName());
+    }
```

Comment:
you can also try using one liner for loops https://stackoverflow.com/questions/40946871/java-one-liner-for-loop

Reply:
The solution on that stackoverflow post is for integers. I think the solution here (https://stackoverflow.com/questions/22649978/java-8-lambda-can-i-generate-a-new-arraylist-of-objects-from-an-intstream) maps integers to objects in order to use the java IntStream interface. Do you think implementing the for loop this way is worthwhile?

Reply:
ya probably leave it like this